File: src/python/base/Protocols/Eth.py
'''
[Narith]
File:   IOManager.py
Author: Saad Talaat
Date:   15th July 2013
brief:  Structure to hold Ethernet info
'''

class Eth():

	# FLAGS
	
	#type of data rep of input
	ISSTRING = False
	
	#packet type flags
	ISIP = False
	ISARP = False

	def __init__(self, dst, src, t):

		# Check if in binary stream format or seperated format
		if(type(dst) == str):
			try:
				assert len(dst.split(":")) == 6 
				assert len(src.split(":")) == 6
				self.__sdst__ = dst
				self.__ssrc__ = src
				self.ISSTRING = True
				self.rawDstSrc()
			except:
				pass

		# if not then assign them to raw variables
		# and conduct string initalization
		self.__dst__ = dst
		self.__src__ = src
		self__type__ = t
		self.srcDstSrc()

		self.__initFlags()


	def __initFlags(self):
		if self.__type__ == '\x08\x06':
			self.ISARP = True
		elif self.__type__ == '\x80\x00':
			self.ISIP = True


	def rawDstSrc(self):
		self.__dst__ = "".join([chr(j) for j in  [int(c,base=16) for c in self.__sdst__.split(":")]])
		self.__src__ = "".join([chr(j) for j in  [int(c,base=16) for c in self.__ssrc__.split(":")]])
		return (self.__dst__, self.__src__)


	#return pair or (dst,src) in seperated representation
	def strDstSrc(self):
		if self.ISSTRING:
			return (self.__sdst__, self.__ssrc__)
		dst = ""
		src = ""
		for c in self.__dst__:
			dst += c.encode('hex')
			dst += ":"
		dst = dst[:len(dst)-1]
		for c in self.__src__:
			src += c.encode('hex')
			src += ":"
		src = src[:len(src)-1]
		self.__sdst__ = dst
		self.__ssrc__ = src
		self.ISSTRING = True
		return (dst,src)

	############################
	# Boolean Packet type checks

	def isIP(self):
		return self.ISIP
	def isARP(self):
		return self.ISARP

File: src/haskell/base/Protocols/Eth.hs
-- [Narith]
-- File: Eth.hs
-- Author: Saad Talaat
-- Date: 17th of July 2013
-- Brief: Structure to hold Ethernet info

module Narith.Protocols.Eth () where
import Narith.Protocol
import Data.ByteString as L
import Data.ByteString.Char8 as C
import Data.List as DL
import Data.Word
-- local datatype holding eth info
data EthType = EthArp | EthIP | EthNone
  deriving (Eq)

-- dst,src can vary from native to formatted representation
data Eth = Eth {
		 dstMac :: String,
		 srcMac :: String,
		 ethType :: EthType } |
	   EthRaw{
		ethRaw :: ByteString
		}

-- Init Ethernet representation
ethInit :: L.ByteString -> Eth
ethInit raw = (Eth dst src (mapType t)) where
  dst = DL.intercalate ":" $DL.map show $L.unpack $L.take 6 raw
  src = DL.intercalate ":" $DL.map show $L.unpack $L.take 6 $snd $L.splitAt 6 raw
  t   = lookup (read (Prelude.concat $DL.map show $L.unpack $L.take 2 $snd $L.splitAt 12 raw) :: Int) ethTypes where
    ethTypes = [(0x8000,EthIP),(0x8006,EthArp)]
  mapType (Just z) = z
  mapType Nothing = EthNone

-- retrieve raw Bytestring representation
ethToRaw :: Eth -> L.ByteString
ethToRaw (Eth dst src t) = L.append d $ L.append s tp where
  d = L.pack (
    DL.map read (
      DL.map C.unpack $(C.splitWith (==':') $ C.pack dst)
    ) :: [Word8])
  s = L.pack (Prelude.map read (Prelude.map C.unpack $(C.splitWith (==':') $ C.pack src)) :: [Word8])
  tp = L.pack $mapType $lookup t ethTypes where
    ethTypes = [(EthIP,[0x80,00]),(EthArp,[0x80,06])]
    mapType (Just z) = z
    mapType Nothing = [00,00]
-- here we suppose a is a string
instance Protocol (Eth) where
  formatProtocol (EthRaw raw) = ethInit raw
  rawProtocol a = EthRaw (ethToRaw a)
